sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala

```diff
@@ -177,7 +177,9 @@ object SQLExecution extends Logging {
             shuffleIds.foreach { shuffleId =>
               queryExecution.shuffleCleanupMode match {
                 case RemoveShuffleFiles =>
-                  SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  // Do not unregister the shuffle on MapOutputTracker here to trigger stage retry.
+                  // Otherwise, downstream tasks will fail with MetadataFetchFailedException.
+                  sc.shuffleDriverComponents.removeShuffle(shuffleId, Utils.isTesting)
```

Comment:
This cleanup is not critical, I think the second parameter `blocking` can always be false

Reply:
The unit tests would be flaky if we use false here.

Reply:
oh I see, can we add some comments to explain it?

Reply:
Sure. Done.